Add `timestamp.parse_rfc3339` function

mooreryan opened the pull request:
See https://github.com/gleam-lang/time/issues/9

Grammar for parsing and for property test generators taken from section [5.6 - Internet Date/Time Format](https://datatracker.ietf.org/doc/html/rfc3339#section-5.6) of RFC 3339.

Most of the testing is done by comparing this to test oracles with property tests.  I have FFI for Erlang and JS versions to compare to as well as a test file of 1000 timestamps parsed by the OCaml ptime library.  That is included because the Erlang and JS versions have slight limitations compared to the the implementation in this module (mainly regarding fractional second precision), but if you don't want it in there I can take it out.

Couple of issues with the tests that will need to be resolved at some point:

- The Erlang oracle used in the property tests will need to be updated once [this](https://github.com/erlang/otp/issues/9279) is fixed upstream.  (The Erlang oracle code adjusts the time to correct for that issue.)
- ~~The `int_uniform` used in `timestamp_rfc3339_timestamp_roundtrip_property_test` doesn't get to the boundaries of `0000` to `9999` years.  It's out of the range of the prng int generator.~~ (I fixed this in a subsequent commit.)
- The round trip test from string -> timestamp -> string will need to be updated once to_rfc3339 handles fractional seconds.

mooreryan commented:
The only part of the interface that I needed to change (other than the parse function) was exposing a `duration.normalised` function so that I could convert a timestamp into a duration.

mooreryan commented:
After fixing the range of the timestamp generator, I had to cherry pick the commit from this PR: https://github.com/gleam-lang/time/pull/10.

lpil commented on test/gleam/time/timestamp_test.gleam:
```diff
@@ -207,3 +214,329 @@ pub fn to_rfc3339_8_test() {
+  parse_rfc3339_should_match_oracle_for("1970-01-01T23:59:60Z")
+}
+
+fn parse_rfc3339_should_match_oracle_for(date_time) {
```
Delete this function and inline the body into all the places it was used please. Tests should be as simple as possible

lpil commented on test/gleam/time/timestamp_test.gleam:
```diff
@@ -207,3 +214,329 @@ pub fn to_rfc3339_8_test() {
+
+  let result = timestamp.parse_rfc3339(date_time)
+  let expected = parse_rfc3339_oracle(date_time)
+  case result, expected {
```
This test is hard too complex and hard to read. Can we filter these out beforehand in the generator please

mooreryan replied:
Yes good point--I think not generating them in the first place, and then using a couple of well chosen example tests will be better here.

lpil commented on src/gleam/time/timestamp.gleam:
````diff
@@ -227,12 +265,388 @@ fn to_civil(minutes: Int) -> #(Int, Int, Int) {
+/// // -> #(1_736_543_370, 0)
+/// 
+/// let assert Error(Nil) = timestamp.parse_rfc3339("1995-10-31")
+/// ```
````
One block per example please, and remove one of the two success cases

lpil commented on src/gleam/time/timestamp.gleam:
```diff
@@ -227,12 +265,388 @@ fn to_civil(minutes: Int) -> #(Int, Int, Int) {
+///   nanosecond precision is tructated.  That is, for the time string, 
+///   `"1970-01-01T00:00:00.1234567899Z"`, the fractional second `.1234567899` 
+///   will be represented as `123_456_789` in the `Timestamp`.
+/// 
```
Do we need to describe the format here? Seems some of this is redundant?

mooreryan replied:
Yes I see what you mean.  These notes are trying to be specific about things that are not really specified in the RFC.  E.g., it says applications may choose to use a space or something else to separate date and time, or that case of T and Z may be considered, and the bit about truncating.

Maybe it would be better in the readme rather than here? Or, if you think it's better, it could be left "unspecified".

lpil commented on src/gleam/time/timestamp.gleam:
```diff
@@ -227,12 +265,388 @@ fn to_civil(minutes: Int) -> #(Int, Int, Int) {
+}
+
+fn do_parse_second_fraction_as_nanoseconds(from bytes, acc acc, pow pow, k k) {
+  case bytes {
```
Collapse this nested case into one and favour > and < over the compare functions please

lpil commented on src/gleam/time/timestamp.gleam:
```diff
@@ -227,12 +265,388 @@ fn to_civil(minutes: Int) -> #(Int, Int, Int) {
+) -> Result(#(Int, BitArray), Nil) {
+  use #(day, bytes) <- result.try(parse_digits(from: bytes, count: 2))
+
+  let is_leap_year = is_leap_year(year)
```
Move this into the clause that uses it please

lpil commented on src/gleam/time/timestamp.gleam:
```diff
@@ -227,12 +265,388 @@ fn to_civil(minutes: Int) -> #(Int, Int, Int) {
+  - 32_045
+}
+
+fn subtract(left: Timestamp, right: Timestamp) -> Timestamp {
```
This function doesn't make sense. You cannot subtract a timestamp from a timestamp.

lpil commented on src/gleam/time/timestamp.gleam:
```diff
@@ -227,12 +265,388 @@ fn to_civil(minutes: Int) -> #(Int, Int, Int) {
+  acc acc: Int,
+  k k: Int,
+) -> Result(#(Int, BitArray), Nil) {
+  case int.compare(k, count) {
```
Collapse these nested cases into one using guards and the int comparison operators please

lpil commented on src/gleam/time/timestamp.gleam:
```diff
@@ -227,12 +265,388 @@ fn to_civil(minutes: Int) -> #(Int, Int, Int) {
+  from bytes: BitArray,
+  count count: Int,
+  acc acc: Int,
+  k k: Int,
```
What's k?

mooreryan replied:
k is just the name of the loop counter. I could change the name to something more descriptive?

lpil reviewed (commented):
Wow! Super nice work!! I've left a bunch of questions and notes inline.

Thank you so much!!

mooreryan commented on test/gleam/time/rfc3339_generator.gleam:
```diff
@@ -0,0 +1,267 @@
+  partial_time <> time_offset
+}
+
+fn full_time_no_secfrac_generator(
```
This generates a timestamp with no "second fraction" part.   Once the `to_rfc3339` function outputs string timestamps with fractional seconds, it can be removed.

("secfrac" is an abbreviation from the RFC 3339 docs for second fraction.  There are some other abbreviations in the tests file that come from that doc.  I could replace them with full names.)

mooreryan replied:
I replaced the abbreviation.

mooreryan commented on src/gleam/time/duration.gleam:
```diff
@@ -24,6 +24,18 @@ pub opaque type Duration {
+/// Ensures the duration is represented with `nanoseconds` being positive and
+/// less than 1 second.
+/// 
+pub fn normalised(
```
I can inline the duration math, which will eliminate the need for this.

mooreryan commented on test/gleam/time/timestamp_test.gleam:
```diff
@@ -207,3 +214,329 @@ pub fn to_rfc3339_8_test() {
+
+pub fn timestamp_rfc3339_timestamp_roundtrip_property_test() {
+  use timestamp <- qcheck.run(
+    config: qcheck.default_config() |> qcheck.with_test_count(10_000),
```
The default test count is 1000 because string generators are pretty slow.  Int generators are very fast though, so I bumped it up to 10,000.  It doesn't have to be--just a way to get more tests in since this generator is fast.

mooreryan replied:
I switched to the default test count.

mooreryan commented:
Thanks for looking at this and the comments!

Most of your comments I was able to address without additional clarification.  I left a couple of follow up comments.  

Would you like me to squash all the tiny commits I created as I was addressing the issues?

mooreryan commented:
Oh two other questions for you:

---

I had thought to make some tests like this:

```gleam
pub fn parse_rfc3339_returns_error_for_bad_timestamp_0_test() {
  timestamp.parse_rfc3339("12349-01-01T00:00:00Z")
  |> should.equal(Error(Nil))
}

pub fn parse_rfc3339_returns_error_for_bad_timestamp_1_test() {
  timestamp.parse_rfc3339("1234-019-01T00:00:00Z")
  |> should.equal(Error(Nil))
}
```

with specific examples that break the grammar in some way, but then I thought it might not be worth the effort.  What do you think? (Ideally, I could just generate non-conforming datetime strings, but that would be a bit of work.)

---

The other thing, do you want me to make a couple of parsing tests that don't use oracles at all?  Some stuff like

```gleam
pub fn blah_test() {
  let assert Ok(ts) = timestamp.parse_rfc3339(...)
  timestamp.to_unix_seconds_and_nanoseconds(ts)
  |> should.equal(#(..., ...))
```

There are quite a few specific examples that are tested (in the `parse_rfc3339_matches_oracle_example_N_test` functions), but they use the oracles, and so you can't glance at the tests to see the expected values.  I don't think this is necessarily an issue, what do you think?

lpil commented:
I don't mind RE tests that show it not parsing. Your call.

> The other thing, do you want me to make a couple of parsing tests that don't use oracles at all

Maybe a couple would be nice! Thank you.

mooreryan commented:
I updated the tests, and made a few more tweaks to address the last unresolved comments. 

The last unresolved issue is regarding how the doc comment of the function should be--whether it should detail the specifications of the parsing (comment is here: https://github.com/gleam-lang/time/pull/12#discussion_r1912156422).  I'm not really sure what is best regarding that comment.

lpil reviewed (approved):
Thank you!!! Fantastic work!

Let's leave it as you've got it there 👍
